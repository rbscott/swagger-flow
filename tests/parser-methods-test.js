import yaml from 'js-yaml';

import parse from '../src/parser';
import {
  DEFAULT_PARSER_CONFIG,
  INLINE_PARSER_CONFIG,
} from './constants';

describe('parse', () => {
  describe('parameters', () => {
    it('Should parse a basic parameter', () => {
      const input = yaml.safeLoad(`
paths:
  /simple/test:
    get:
      parameters:
      - name: param1
        in: query
        type: string
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should parse multiple parameters', () => {
      const input = yaml.safeLoad(`
paths:
  /simple/test:
    get:
      parameters:
      - name: param1
        in: query
        type: string
      - name: param2
        in: query
        type: boolean
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should parse global parameters', () => {
      const input = yaml.safeLoad(`
parameters:
- name: globalParam1
  in: query
  type: string

paths:
  /simple/test:
    get:
      parameters:
      - name: param1
        in: query
        type: string
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should parse path parameters', () => {
      const input = yaml.safeLoad(`
paths:
  /simple/test:
    parameters:
    - name: pathParam1
      in: query
      type: string
    get:
      parameters:
      - name: param1
        in: query
        type: string
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should inherit properly', () => {
      const input = yaml.safeLoad(`
parameters:
- name: sameParam
  in: query
  type: string

paths:
  /simple/test:
    parameters:
    - name: sameParam
      in: query
      type: boolean
    get:
      parameters:
      - name: sameParam
        in: query
        type: number
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should use the operationId without a controller', () => {
      const input = yaml.safeLoad(`
parameters:
paths:
  /simple/test:
    parameters:
    - name: sameParam
      in: query
      type: boolean
    get:
      operationId: Operation
      parameters:
      - name: sameParam
        in: query
        type: number
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should use the operationId with a path controller', () => {
      const input = yaml.safeLoad(`
parameters:
paths:
  /simple/test:
    x-flow-controller: Controller
    parameters:
    - name: sameParam
      in: query
      type: boolean
    get:
      operationId: Operation
      parameters:
      - name: sameParam
        in: query
        type: number
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should use the operationId with a method controller', () => {
      const input = yaml.safeLoad(`
parameters:
paths:
  /simple/test:
    x-flow-controller: Controller
    parameters:
    - name: sameParam
      in: query
      type: boolean
    get:
      x-flow-controller: Method
      operationId: Operation
      parameters:
      - name: sameParam
        in: query
        type: number
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should split out the body if requested.', () => {
      const input = yaml.safeLoad(`
parameters:
paths:
  /simple/test:
    get:
      parameters:
      - name: body
        in: body
        type: number
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should split out the body with a rich schema.', () => {
      const input = yaml.safeLoad(`
parameters:
paths:
  /simple/test:
    get:
      parameters:
      - name: body
        in: body
        schema:
          type: object
          properties:
            - name: param1
              type: string
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should inline body if requested.', () => {
      const input = yaml.safeLoad(`
parameters:
paths:
  /simple/test:
    get:
      parameters:
      - name: body
        in: body
        type: number
`);
      expect(parse(input, INLINE_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should inline the body with a rich schema.', () => {
      const input = yaml.safeLoad(`
parameters:
paths:
  /simple/test:
    get:
      parameters:
      - name: body
        in: body
        schema:
          type: object
          properties:
            - name: param1
              type: string
`);
      expect(parse(input, INLINE_PARSER_CONFIG).items).toMatchSnapshot();
    });
  });

  describe('response', () => {
    it('Should parse a basic response', () => {
      const input = yaml.safeLoad(`
paths:
  /simple/test:
    get:
      responses:
        200:
          schema:
            type: object
            properties:
              resp1:
                type: string
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should parse an operationId without a controller', () => {
      const input = yaml.safeLoad(`
paths:
  /simple/test:
    get:
      operationId: Operation
      responses:
        200:
          schema:
            type: object
            properties:
              resp1:
                type: string
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should parse an operationId with a path controller', () => {
      const input = yaml.safeLoad(`
paths:
  /simple/test:
    x-flow-controller: Controller
    get:
      operationId: Operation
      responses:
        200:
          schema:
            type: object
            properties:
              resp1:
                type: string
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });

    it('Should parse an operationId with a method controller', () => {
      const input = yaml.safeLoad(`
paths:
  /simple/test:
    x-flow-controller: Controller
    get:
      x-flow-controller: Method
      operationId: Operation
      responses:
        200:
          schema:
            type: object
            properties:
              resp1:
                type: string
`);
      expect(parse(input, DEFAULT_PARSER_CONFIG).items).toMatchSnapshot();
    });


  });

});
