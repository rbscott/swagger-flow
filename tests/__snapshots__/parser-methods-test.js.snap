// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse parameters Should inherit properly 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [
      Array [
        "sameParam",
        Object {
          "id": "#/simple/test/get/sameParam",
          "name": "RqstSimpleTestGet_Field_sameParam",
          "type": "integer",
        },
      ],
    ],
    "id": "#/simple/test/get",
    "name": "RqstSimpleTestGet",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse parameters Should inline body if requested. 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [
      Array [
        "body",
        Object {
          "id": "#/simple/test/get/body",
          "name": "RqstSimpleTestGet_Field_body",
          "type": "integer",
        },
      ],
    ],
    "id": "#/simple/test/get",
    "name": "RqstSimpleTestGet",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse parameters Should inline the body with a rich schema. 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [
      Array [
        "body",
        Object {
          "additionalProperties": undefined,
          "fields": Array [
            Array [
              "0",
              Object {
                "id": "#/simple/test/get/body/Field_0",
                "name": "RqstSimpleTestGet_Field_body_Field_0",
                "type": "string",
              },
            ],
          ],
          "id": "#/simple/test/get/body",
          "name": "RqstSimpleTestGet_Field_body",
          "required": Array [],
          "type": "object",
        },
      ],
    ],
    "id": "#/simple/test/get",
    "name": "RqstSimpleTestGet",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse parameters Should parse a basic parameter 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [
      Array [
        "param1",
        Object {
          "id": "#/simple/test/get/param1",
          "name": "RqstSimpleTestGet_Field_param1",
          "type": "string",
        },
      ],
    ],
    "id": "#/simple/test/get",
    "name": "RqstSimpleTestGet",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse parameters Should parse global parameters 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [
      Array [
        "globalParam1",
        Object {
          "id": "#/parameters/globalParam1",
          "name": "parameters_Field_globalParam1",
          "type": "string",
        },
      ],
      Array [
        "param1",
        Object {
          "id": "#/simple/test/get/param1",
          "name": "RqstSimpleTestGet_Field_param1",
          "type": "string",
        },
      ],
    ],
    "id": "#/simple/test/get",
    "name": "RqstSimpleTestGet",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse parameters Should parse multiple parameters 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [
      Array [
        "param1",
        Object {
          "id": "#/simple/test/get/param1",
          "name": "RqstSimpleTestGet_Field_param1",
          "type": "string",
        },
      ],
      Array [
        "param2",
        Object {
          "id": "#/simple/test/get/param2",
          "name": "RqstSimpleTestGet_Field_param2",
          "type": "boolean",
        },
      ],
    ],
    "id": "#/simple/test/get",
    "name": "RqstSimpleTestGet",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse parameters Should parse path parameters 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [
      Array [
        "pathParam1",
        Object {
          "id": "#/simple/test/pathParam1",
          "name": "SimpleTest/_Field_pathParam1",
          "type": "string",
        },
      ],
      Array [
        "param1",
        Object {
          "id": "#/simple/test/get/param1",
          "name": "RqstSimpleTestGet_Field_param1",
          "type": "string",
        },
      ],
    ],
    "id": "#/simple/test/get",
    "name": "RqstSimpleTestGet",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse parameters Should split out the body if requested. 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [],
    "id": "#/simple/test/get",
    "name": "RqstSimpleTestGet",
    "required": Array [],
    "type": "object",
  },
  Object {
    "id": "#/simple/test/get/body",
    "name": "RqstSimpleTestGetBody",
    "type": "integer",
  },
]
`;

exports[`parse parameters Should split out the body with a rich schema. 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [],
    "id": "#/simple/test/get",
    "name": "RqstSimpleTestGet",
    "required": Array [],
    "type": "object",
  },
  Object {
    "additionalProperties": undefined,
    "fields": Array [
      Array [
        "0",
        Object {
          "id": "#/simple/test/get/body/Field_0",
          "name": "RqstSimpleTestGetBody_Field_0",
          "type": "string",
        },
      ],
    ],
    "id": "#/simple/test/get/body",
    "name": "RqstSimpleTestGetBody",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse parameters Should use the operationId with a method controller 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [
      Array [
        "sameParam",
        Object {
          "id": "#/simple/test/get/sameParam",
          "name": "RqstMethodOperation_Field_sameParam",
          "type": "integer",
        },
      ],
    ],
    "id": "#/simple/test/get",
    "name": "RqstMethodOperation",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse parameters Should use the operationId with a path controller 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [
      Array [
        "sameParam",
        Object {
          "id": "#/simple/test/get/sameParam",
          "name": "RqstControllerOperation_Field_sameParam",
          "type": "integer",
        },
      ],
    ],
    "id": "#/simple/test/get",
    "name": "RqstControllerOperation",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse parameters Should use the operationId without a controller 1`] = `
Array [
  Object {
    "additionalProperties": false,
    "fields": Array [
      Array [
        "sameParam",
        Object {
          "id": "#/simple/test/get/sameParam",
          "name": "RqstOperation_Field_sameParam",
          "type": "integer",
        },
      ],
    ],
    "id": "#/simple/test/get",
    "name": "RqstOperation",
    "required": Array [],
    "type": "object",
  },
]
`;

exports[`parse response Should parse a basic response 1`] = `
Array [
  Object {
    "additionalProperties": undefined,
    "fields": Array [
      Array [
        "resp1",
        Object {
          "id": "#/simple/test/get/200/Field_resp1",
          "name": "RespSimpleTestGet200_Field_resp1",
          "type": "string",
        },
      ],
    ],
    "id": "#/simple/test/get/200",
    "name": "RespSimpleTestGet200",
    "required": Array [],
    "type": "object",
  },
  Object {
    "choices": Array [
      Object {
        "id": "/simple/test/choices/RespSimpleTestGet200",
        "name": "RespSimpleTestGet_Ref_RespSimpleTestGet200",
        "ref": "#/simple/test/get/200",
        "type": "ref",
      },
    ],
    "id": "#/simple/test/get",
    "name": "RespSimpleTestGet",
    "required": Array [],
    "type": "anyOf",
  },
]
`;

exports[`parse response Should parse an operationId with a method controller 1`] = `
Array [
  Object {
    "additionalProperties": undefined,
    "fields": Array [
      Array [
        "resp1",
        Object {
          "id": "#/simple/test/get/200/Field_resp1",
          "name": "RespMethodOperation200_Field_resp1",
          "type": "string",
        },
      ],
    ],
    "id": "#/simple/test/get/200",
    "name": "RespMethodOperation200",
    "required": Array [],
    "type": "object",
  },
  Object {
    "choices": Array [
      Object {
        "id": "/simple/test/choices/RespMethodOperation200",
        "name": "RespMethodOperation_Ref_RespMethodOperation200",
        "ref": "#/simple/test/get/200",
        "type": "ref",
      },
    ],
    "id": "#/simple/test/get",
    "name": "RespMethodOperation",
    "required": Array [],
    "type": "anyOf",
  },
]
`;

exports[`parse response Should parse an operationId with a path controller 1`] = `
Array [
  Object {
    "additionalProperties": undefined,
    "fields": Array [
      Array [
        "resp1",
        Object {
          "id": "#/simple/test/get/200/Field_resp1",
          "name": "RespControllerOperation200_Field_resp1",
          "type": "string",
        },
      ],
    ],
    "id": "#/simple/test/get/200",
    "name": "RespControllerOperation200",
    "required": Array [],
    "type": "object",
  },
  Object {
    "choices": Array [
      Object {
        "id": "/simple/test/choices/RespControllerOperation200",
        "name": "RespControllerOperation_Ref_RespControllerOperation200",
        "ref": "#/simple/test/get/200",
        "type": "ref",
      },
    ],
    "id": "#/simple/test/get",
    "name": "RespControllerOperation",
    "required": Array [],
    "type": "anyOf",
  },
]
`;

exports[`parse response Should parse an operationId without a controller 1`] = `
Array [
  Object {
    "additionalProperties": undefined,
    "fields": Array [
      Array [
        "resp1",
        Object {
          "id": "#/simple/test/get/200/Field_resp1",
          "name": "RespOperation200_Field_resp1",
          "type": "string",
        },
      ],
    ],
    "id": "#/simple/test/get/200",
    "name": "RespOperation200",
    "required": Array [],
    "type": "object",
  },
  Object {
    "choices": Array [
      Object {
        "id": "/simple/test/choices/RespOperation200",
        "name": "RespOperation_Ref_RespOperation200",
        "ref": "#/simple/test/get/200",
        "type": "ref",
      },
    ],
    "id": "#/simple/test/get",
    "name": "RespOperation",
    "required": Array [],
    "type": "anyOf",
  },
]
`;
